pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -1060,7 +1060,31 @@ func isKernelPid(pid int) bool {
 	return err != nil && os.IsNotExist(err)
 }
 
+// GetCapacity returns node capacity data for "cpu", "memory", "ephemeral-storage", and "huge-pages*"
+// At present this method is only invoked when introspecting ephemeral storage
 func (cm *containerManagerImpl) GetCapacity() v1.ResourceList {
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LocalStorageCapacityIsolation) {
+		// We store allocatable ephemeral-storage in the capacity property once we Start() the container manager
+		if _, ok := cm.capacity[v1.ResourceEphemeralStorage]; !ok {
+			// If we haven't yet stored the capacity for ephemeral-storage, we can try to fetch it directly from cAdvisor,
+			if cm.cadvisorInterface != nil {
+				rootfs, err := cm.cadvisorInterface.RootFsInfo()
+				if err != nil {
+					klog.ErrorS(err, "Unable to get rootfs data from cAdvisor interface")
+					// If the rootfsinfo retrieval from cAdvisor fails for any reason, fallback to returning the capacity property with no ephemeral storage data
+					return cm.capacity
+				}
+				// We don't want to mutate cm.capacity here so we'll manually construct a v1.ResourceList from it,
+				// and add ephemeral-storage
+				capacityWithEphemeralStorage := v1.ResourceList{}
+				for rName, rQuant := range cm.capacity {
+					capacityWithEphemeralStorage[rName] = rQuant
+				}
+				capacityWithEphemeralStorage[v1.ResourceEphemeralStorage] = cadvisor.EphemeralStorageCapacityFromFsInfo(rootfs)[v1.ResourceEphemeralStorage]
+				return capacityWithEphemeralStorage
+			}
+		}
+	}
```

Comment:
Before, to read cm.capacity, there is a read lock. The lock is no longer needed?

Reply:
where is that read lock? I'm not aware of it

Reply:
not directly related to your change, but is it possible that cm.capacity is set while at the same time it is being retrieved?

Reply:
> where is that read lock? I'm not aware of it

https://github.com/kubernetes/kubernetes/pull/48567#diff-c07d343a8d459721656cf7ddacda71506bde01a5cac288d15a4aba6bfc9a1cae

it got removed from here https://github.com/kubernetes/kubernetes/pull/59769/files

Reply:
That is certainly possible (this change would definitely not change that behavior if that's permitted). The important thing is that this change doesn't mutate `cm.capacity` in any way. If the `capacity` field is not present at the time that this `GetCapacity` method is invoked, then we attempt (best-effort) to retrieve the storage capacity of the root filesystem manually, and if we are able, we return *that*.

Thanks for the review, hope that makes sense!

Reply:
the last thing I want to confirm retry is not needed here? I still see some value of retrying, especially if no inconsistent data will be returned from cadvisor?
What is the concern of retry?

Reply:
If you read the prior thread history (I know it's long!) the concern was the possible startup delay added by doing retries in this boostrap code surface area, and any undesirable side-effects that may occur. Doing a cAdvisor lookup in this particular place is already imposing a very tiny delay, and any more was seen as too much risk.

The above was the review consensus as far as I interpreted it. I originally advocated for retries, but my tests without the retry proved that the particular symptom I was trying to fix was equivalently addresses without any retries.